File: backend/services/liquidacion_service.py
from sqlalchemy.orm import Session, joinedload
from sqlalchemy.exc import IntegrityError
from fastapi import HTTPException
from models.liquidacion import Liquidacion
from schemas.liquidacion_schema import LiquidacionCreate, LiquidacionUpdate


def create_liquidacion(db: Session, liquidacion: LiquidacionCreate):
    try:
        db_liquidacion = Liquidacion(
            id_contrato=liquidacion.id_contrato,
            fecha_liquidacion=liquidacion.fecha_liquidacion,
            id_motivo_terminacion=liquidacion.id_motivo_terminacion,
            total_liquidacion=liquidacion.total_liquidacion
        )

        db.add(db_liquidacion)
        db.commit()
        db.refresh(db_liquidacion)

# 🔹 Traer la liquidación con relaciones completas
        liquidacion_full = (
            db.query(Liquidacion)
            .options(
                joinedload(Liquidacion.motivo_terminacion),
                joinedload(Liquidacion.detalles_liquidacion)
            )
            .filter(Liquidacion.id_liquidacion == db_liquidacion.id_liquidacion)
            .first()
        )

        return liquidacion_full


    except IntegrityError:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail="Error al crear liquidación (validación o FK no válida)"
        )
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e))


def get_all_liquidaciones(db: Session):
    return (
        db.query(Liquidacion)
        .options(
            joinedload(Liquidacion.motivo_terminacion),
            joinedload(Liquidacion.detalles_liquidacion)
        )
        .all()
    )


def get_liquidacion_by_id(db: Session, id_liquidacion: int):
    liquidacion = (
        db.query(Liquidacion)
        .options(
            joinedload(Liquidacion.motivo_terminacion),
            joinedload(Liquidacion.detalles_liquidacion)
        )
        .filter(Liquidacion.id_liquidacion == id_liquidacion)
        .first()
    )

    if not liquidacion:
        raise HTTPException(status_code=404, detail="Liquidación no encontrada")

    return liquidacion


def update_liquidacion(db: Session, id_liquidacion: int, liquidacion_update: LiquidacionUpdate):
    try:
        liquidacion_db = get_liquidacion_by_id(db, id_liquidacion)

        for key, value in liquidacion_update.dict(exclude_unset=True).items():
            setattr(liquidacion_db, key, value)

        db.commit()
        db.refresh(liquidacion_db)
        return liquidacion_db

    except IntegrityError:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail="Error al actualizar liquidación (validación o FK no válida)"
        )
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e))


def delete_liquidacion(db: Session, id_liquidacion: int):
    try:
        liquidacion_db = get_liquidacion_by_id(db, id_liquidacion)
        db.delete(liquidacion_db)
        db.commit()
        return {"message": "Liquidación eliminada correctamente"}

    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/services/LiquidacionService.js
import api from "./api";

export const createLiquidacion = async (data) => {
  try {
    const response = await api.post("/liquidacion/", data);
    return { ok: true, data: response.data };
  } catch (error) {
    console.error("❌ Error al crear liquidación:", error);
    return {
      ok: false,
      message: error.response?.data?.detail || "Error al crear liquidación",
    };
  }
};

export const getAllLiquidaciones = async () => {
  try {
    const response = await api.get("/liquidacion/");
    return { ok: true, data: response.data };
  } catch (error) {
    console.error("❌ Error al obtener liquidaciones:", error);
    return {
      ok: false,
      message: error.response?.data?.detail || "Error al obtener liquidaciones",
    };
  }
};

export const getLiquidacionById = async (id_liquidacion) => {
  try {
    const response = await api.get(`/liquidacion/${id_liquidacion}`);
    return { ok: true, data: response.data };
  } catch (error) {
    console.error("❌ Error al obtener liquidación:", error);
    return {
      ok: false,
      message: error.response?.data?.detail || "Error al obtener liquidación",
    };
  }
};

export const updateLiquidacion = async (id_liquidacion, data) => {
  try {
    const response = await api.put(`/liquidacion/${id_liquidacion}`, data);
    return { ok: true, data: response.data };
  } catch (error) {
    console.error("❌ Error al actualizar liquidación:", error);
    return {
      ok: false,
      message: error.response?.data?.detail || "Error al actualizar liquidación",
    };
  }
};

export const deleteLiquidacion = async (id_liquidacion) => {
  try {
    await api.delete(`/liquidacion/${id_liquidacion}`);
    return { ok: true };
  } catch (error) {
    console.error("❌ Error al eliminar liquidación:", error);
    return {
      ok: false,
      message: error.response?.data?.detail || "Error al eliminar liquidación",
    };
  }
};
